message.go

```diff
@@ -46,30 +44,31 @@ func (m *Message) SetData(data interface{}) error {
 		return fmt.Errorf("failed to set data as struct is expected, got: %s", reflect.TypeOf(str).Kind())
 	}
 
-	for i, fl := range m.fields {
+	for i, specField := range m.fields {
 		fieldName := fmt.Sprintf("F%d", i)
 
 		// get the struct field
 		dataField := str.FieldByName(fieldName)
 
-		if dataField == (reflect.Value{}) || dataField.IsNil() {
+		// no data field was found with this name
+		if dataField == (reflect.Value{}) {
 			continue
 		}
 
-		if dataField.Type() != reflect.TypeOf(fl) {
-			return fmt.Errorf("failed to set data: type of the field %d: %v does not match the type in the spec: %v", i, dataField.Type(), reflect.TypeOf(fl))
+		isNil := false
+		if dataField.IsNil() {
+			isNil = true
+			dataField.Set(reflect.New(dataField.Type().Elem()))
+		}
+		if err := specField.SetData(dataField.Interface()); err != nil {
+			return fmt.Errorf("failed to set data for field %d: %w", i, err)
+		}
+		if isNil {
+			continue
 		}
 
-		// set data field spec for the message spec field
-		specField := m.fields[i]
-		df := dataField.Interface().(field.Field)
-		df.SetSpec(specField.Spec())
-
-		// use data field as a message field
-		m.fields[i] = df
 		m.fieldsMap[i] = struct{}{}
```

Comment:
suggestion not working because of common lines, but this:
```
		if isNil {
			continue
		}
		m.fieldsMap[i] = struct{}{}
```
to
```
		if !isNil {
			m.fieldsMap[i] = struct{}{}
		}
```